### ISSUE TITLE
Broken on new GitHub UI beta

### ISSUE THREAD

**Issue by fregante** [OctoLinker/OctoLinker#1695]
### Expected Behavior

<img width="486" alt="Screen Shot 2" src="https://user-images.githubusercontent.com/1402241/201465175-9831cf5a-ffdb-4cc7-96ff-ce57463ad25d.png">

### What actually happened?

<img width="471" alt="Screen Shot 1" src="https://user-images.githubusercontent.com/1402241/201465176-4c6e8c36-85bc-4dc7-9f94-2fbd2ee9f7ea.png">

### URL

https://github.com/sindresorhus/linkify-issues/blob/main/index.js

### Anything else we should know?

See https://github.blog/changelog/2022-11-09-introducing-an-all-new-code-search-and-code-browsing-experience/#the-all-new-code-browsing-experience

**Comment by fregante** [comment]
I see there's already a PR open 💙

- https://github.com/OctoLinker/OctoLinker/pull/1694

**Comment by xt0rted** [comment]
6.10.5 is in the chrome web store now. This gets OctoLinker working again but the new site is adding extra classes to the elements we're adding links to, so on hover they're styled just like GitHub's code navigation.

<img src="https://user-images.githubusercontent.com/831974/201543797-e827d599-c6f5-44f2-aea1-655a0d217ddf.png">

**Comment by stefanbuck** [comment]
Yeah, I've noticed the same but decided to ignore it for now. However, I just noticed another issue that turbo-link don't re-invoke OctoLinker anymore.

**Comment by xt0rted** [comment]
Was just about to say the same. I have a fix for the link styling but not for triggering us to run on page change.

**Comment by Araxeus** [comment]
Is it just me or does the extension not work at all?
doesn't seem like the PR fixed anything (or they changed the DOM structure again?)

**Comment by fregante** [comment]
What browser are you using?

**Comment by Araxeus** [comment]
> What browser are you using?

I tried on Chrome, Firefox, and Firefox nightly

(I'm not sure when did it break, I only know for sure that I noticed today that it doesn't work)

**Comment by jdreesen** [comment]
Doesn't work for me either with latest Firefox.

**Comment by Araxeus** [comment]
I just noticed that it works inside codeblocks that are in markdown files… 😅

![image](https://user-images.githubusercontent.com/78568641/212760350-e5e892e3-7a9a-478a-bd4f-4e1ec94d1654.png)

but not in actual code files
![image](https://user-images.githubusercontent.com/78568641/212760440-9134bafc-03d2-4978-baa9-e3e209ed6d89.png)

**Comment by fregante** [comment]
There's a bigger issue now: GitHub started using an overlaid `textarea` 🙄 so even if you linkify the dom you can't click it. Maybe a `z-index` on the links will work, otherwise you'll have to rewrite the replacer to use GitHub’s way of highlighting symbols instead.

Related:

- https://github.com/refined-github/refined-github/issues/6336#issuecomment-1498645639

**Comment by silverwind** [comment]
Extension still works fine on [old ui](https://github.com/orgs/community/discussions/54546#discussioncomment-5891649) 😆.

**Comment by silverwind** [comment]
> Extension still works fine on [old ui](https://github.com/orgs/community/discussions/54546#discussioncomment-5891649) 😆.

No longer working as of today, GH is forcing the shitty ui onto logged out users well.

**Comment by stefanbuck** [comment]
I tried to make OctoLinker work with the new UI, but due to the fundamental changes they made, getting OctoLinker working is quite challenging

**Comment by jsumners** [comment]
@stefanbuck is your estimation that this extension is no longer viable?

**Comment by fregante** [comment]
It's a hard issue and it requires a rewrite of the parser/linkifier. If anyone comes up with a practical solution or PR it can definitely help. The main issue is basically lack of time to explore possible solutions.

**Comment by texastoland** [comment]
Just an update that the new UI is now default and can no longer be disabled.

**Comment by cxhello** [comment]
It doesn't work when I use chrome browser.

**Comment by yf-hk** [comment]
Is it working now?

**Comment by texastoland** [comment]
Someone is hiding every comment. The last commit human commit 03b5e9e was in 2022. It was an incredibly useful extension while it existed. Due to no fault of the maintainers my impression is it would be convoluted to make work again. Consider abandoned for now 🎖️

**Comment by jsumners** [comment]
They are being hidden because they are not providing anything useful to the issue. If the issue is open, then it isn't resolved. Asking "still broken?" helps no one. If you feel strongly about the issue, submit a pull request to address it.

**Comment by fregante** [comment]
> Someone is hiding every comment

It's me. As @jsumners said, "still broken" and "broken in my browser" doesn't add any information that isn't already here.

> it would be convoluted to make work again

Yes, this has been the state since my May 2023 comment above: https://github.com/OctoLinker/OctoLinker/issues/1695#issuecomment-1534099619

Octolinker still works in some views, like readmes, comments, commits and PRs, but not on the React-based code viewer. This view has been a major pain for all GitHub extensions, e.g. 

- https://github.com/refined-github/refined-github/issues/6554

**Comment by xt0rted** [comment]
Back when the new React UI was announced, @maximedegreve asked me on Twitter what we'd need to keep this and other extensions working. At the time #1694 got us going again, but now that a lot more has changed and broken this even more is there anything they might be able to do to help?

**Comment by fregante** [comment]
I think the only thing they could do is open up their symbols UI. If you can add symbols, then they would take care of making them clickable

**Comment by maximedegreve** [comment]
Hey @xt0rted 

I will forward this internally. I can't promise anything though. 
Thanks for bringing this to our attention.

**Comment by AdamShwert** [comment]
Hi folks, I'm from the repos team, and I have some ideas for how it could work for you. We unfortunately make the visible code lines that you see un-interactable for a variety of reasons not worth going into here, but there should be a way to make it work relatively easily. Underneath the visible code lines that you see is an invisible text area which contains the full body text of a given file. When a user clicks within the code lines, their cursor position within the text area is being updated to represent where they clicked. The text area, which has the id of `read-only-cursor-text-area` (and we will not be changing it), should be easily queryable by an extension to grab the `selectionStart` and `selectionEnd`. Adding an extra on click handler to the text area should be able to capture any potential clicks to do with what you wish. The `selectionStart` and selectionEnd` are just character offsets within the value of the text area (which includes newlines).

I'm not sure how you all were determining which paths should be clickable, but assuming it was some sort of text scraping, you should be able to grab the value of the text area (which would be the entire body of the file) and see if the `selectionStart` and `selectionEnd` of the text area (which are the same if it is not a highlight) lined up with what should be a clickable path. Getting the value of the text area and precomputing which ranges of character offsets line up with file paths to match against `selectionStart` when that changes should be relatively easy to do. The hard part you likely already did and can re-use, which would be determining where that path should resolve to.

**Comment by jsumners** [comment]
> Underneath the visible code lines that you see is an invisible text area which contains the full body text of a given file. When a user clicks within the code lines, their cursor position within the text area is being updated to represent where they clicked.

Off-topic: but I think this finally explains why all of a sudden clicking anywhere in a code view breaks keyboard history navigation (i.e. it breaks cmd+backspace to navigate backward). I would fervently welcome a setting that disables this.

**Comment by fregante** [comment]
@AdamShwert thank you for the input! I think two parts are still hard to achieve:

- showing that a specific piece of text is clickable: the last time I checked, the component used :before instead of regular text nodes, so splitting and linkifying a part of that wasn't straightforward
- showing a hover state: with your suggestion I'm not sure it's possible/simple to find the exact word under the cursor via `mousemove`

I think that if the first part can be done, we can just wrap it in a proper link and bring it up via z-index.

While you're here, could GitHub natively do what OctoLinker does? It doesn't look substantially different from the cross-file symbol detection that GitHub already has, at least for _some_ of the formats OctoLinker supports.

**Comment by xt0rted** [comment]
I ran into a few issues adding links to the rendered source:

1. The `textarea` has a z-index of 1 so anything in the visible text remains under the `textarea` no matter what z-index you give it
2. Wrapping the `span` with `<a>` like OctoLinker currently does works fine and you get a link, but you can't click on it and there's no hover state (maybe this could be done with js?)
3. The visible text has `pointer-events: none` which prevents clicking on any of the links that are added when you start changing z-indexes

The `textarea` was added for accessibility and so you can navigate/copy the source with just the keyboard. I'm sure there's others, but those were the two main ones I was told.

In all my tests the only way I see to make OctoLinker work is to change the z-indexes and then accessibility/keyboard navigation goes out the window. Tabbing through the page/code may be awkward since you have to go through to code in the `textarea` before you wrap back up to the top of the rendered code and start going through the links. The other behavior issue was I stopped being able to interact with the `textarea` completely so I couldn't copy any text even with a mouse.

Since the raw text of the file is on the page now that actually simplifies how we parse it (no longer have to hit the api in this view, but still would elsewhere). The issues I see are all in displaying a link that you can interact with.

**Comment by AdamShwert** [comment]
> While you're here, could GitHub natively do what OctoLinker does? It doesn't look substantially different from the cross-file symbol detection that GitHub already has, at least for _some_ of the formats OctoLinker supports.

We're discussing doing this, so stay tuned on that front! If we did end up implementing it, we would be doing it without any sort of hover state or anything of the sort, similar to how symbols work now. The issue for making it standard behavior is that people wouldn't necessarily expect clicking on the import to work in that way with no hover state, so we'd need to figure out a good way to let people know that clicking on imports will open a new tab. 


As for making it appear as a link, I think currently it would be very hard to do unfortunately. We have some changes in the works along with firefox/chromium, which when chromium 124 comes out later this month, should make it considerably easier to at least apply the underline. We're doing the `::before` stuff now as you mention, but that will not be necessary for users with firefox and chromium versions above 124. We will be using the `inert` property which hides content from browser find and also makes it so mouse events don't occur on any `inert` segments. This means hover state still would not be possible, but wrapping it in an `<a>` tag so that it is underlined should be much easier. 

You could probably do something extremely hacky and listen on every mouse move and determine which code line the user is hovering over by using scroll offsets and mouse offsets, but with various features such as text only zoom this would likely be very brittle. It would also be a big performance drag.

**Comment by fregante** [comment]
> clicking on imports will open a new tab

I hope you mean "clicking on imports will change the file" since almost no links on GitHub.com open new tabs (also, [nngroup article](https://www.nngroup.com/articles/new-browser-windows-and-tabs/))

In any case it would be very helpful because, as you say, linkifying the current editor will continue to be tough for extensions.

**Comment by AdamShwert** [comment]
> > clicking on imports will open a new tab
> 
> I hope you mean "clicking on imports will change the file" since almost no links on GitHub.com open new tabs (also, [nngroup article](https://www.nngroup.com/articles/new-browser-windows-and-tabs/))

Yeah, we haven't discussed anything yet so that probably would have come up during our discussions. Either way, we'd need a way to let people know that they're going to replace the file they are currently looking at by clicking. An underline might not be enough without the hover state being there too.

**Comment by JuroOravec** [comment]
Hi, I might look into this in the next few days. Yesterday I upgraded Chrome after a long time, and found that the extension is no longer available (which is a shame, I liked it a lot). I was feeling beefy after improving AlpineJS perf by 30% over the last weekend, and I was like "no way this cannot work, at least until Github uses canvas for the code!".

Now, to be frank, I think OctoLinker should and will be on its way out eventually - I think that once you get to the point where you REALLY need to navigate a codebase and get to understand it, it's so much easier to open the codespace (remote IDE).

But on the other hand, when you need to just navigate a few files, OctoLinker is nice, still.

So I looked into it and got to the same approach as @AdamShwert [above](https://github.com/OctoLinker/OctoLinker/issues/1695#issuecomment-2045505608).

Altho I don't entirely agree with the following:

> You could probably do something extremely hacky and listen on every mouse move and determine which code line the user is hovering over by using scroll offsets and mouse offsets, but with various features such as text only zoom this would likely be very brittle. It would also be a big performance drag.

While I agree making links work WILL be hacky, I don't think that it necessarily should be brittle:

- Correct me if wrong, but my guess is that any kind of zoom (text only or otherwise) should preserve HTML element boundaries, no? If a text scale up, and it's wrapped in `<span>`, then the boundary of that `<span>` should get larger too.

- As for the performance, I agree. But at the same time, I think we can get away with e.g. querying the cursor position only every 50ms. Still would look instanteneous, but would avoid lots of re-computations.

Below is some additional thoughts I had along the way:

## Why `https://octolinker-api.now.sh/`?

If I understand [the API repo](https://github.com/OctoLinker/api) correctly, it was used to find a Github URL for given package.

With regards to https://github.com/OctoLinker/OctoLinker/issues/1907, I think, in order to keep the extension low-maintenance, that this feature isn't needed. Instead, the links to 3rd party packages could easily just point to respective registries like NPM, etc.

Again, I may be wrong, but I assume that without the API, all that the extension needs is just hosting for a static site, which can be e.g. under Github for free.

## Robustness - versioning

As a newcomer to this codebase, and given that the project hasn't been updated in a while, I find it hard to work with the logic  that finds and selects the elements on the page. Because I don't know which edge cases were for old versions, and whether those old versions will ever be relevant, etc.

So while I'm at it, I also suggest to update the project structure so that the selectors (and plugins) are versioned (e.g. think of API or data versioning policies).

Practically, it would look like this:
1. There would be a directory of versioned implementations.
2. Individual implementaitons (e.g. files or directories) could be named like `2025-08-01`, or more detailed like `2025-08-01--mobile`.
3. Each such implementation would define some checks, e.g. "if there is `github-class` on this page, it's this version".
4. Implementations would be ordered from the latest to oldest, and the first one that passes the checks would be chosen.

```
|- versions
   |- 2025-08-01/
      |- index.js
      |- plugins/
         |- index.js       // Contains all supported plugins
         |- javascript/ 
            |- index.js
            ...
         |- typescript/
   |- 2025-08-01--mobile
   |- 2024-07-30
```

Plugins would be versioned too, because it can totally happen that one day, everything about Github UI would remain the same, but only some CSS classes for e.g. Rust files would be changed. And so the versioning model should be able to handle that.

In code this could look like so:

```js
// 2025-08-01/index.js

import { default as prevVersion } from '../2024-07-30';

export default const {
    version: "2025-08-01",
    impl: {
        // Copy everything from previous version
        ...prevVersion.impl,

       plugins: {
         ...prevVersion.impl.plugins,

          // But define new selectors for Rust files
         rust: {
           getPatterns() {
             ...
           }
         }
       },
    }
}
```

## Robustness - URL-based flows

Another part of the issue of navigating the current codebase is that it's trying to cover all cases at the same time - E.g. when processing the "blob" element, it tries 3-4 different ways to process it. And I don't know which approach corresponds to what page and what starting HTML.

On the other hand, when you navigate Github, it's quite clear to see that there's a limited number of cases that need to be handled (below). So I suggest to define the logic around these views - each versioned implementation would define functions for how to handle individual views.

So again, if the Github UI changed only e.g. on the "compare" page, a new "implementation" would be created, which would reuse all the code from the previous version, except for the function that defines e.g. how to find the "blobs" on the Compare page.

To choose between these views, I think it's best to rely on the URL. I believe that URLs are less likely to change over the lifetime of the app, because end users can share them, and so Github wouldn't want to go breaking those URLs.

   1. FILE VIEW
      - IDentify by having `/tree/` URL
        - either points to the latest commit of given branch
          https://github.com/OctoLinker/OctoLinker/tree/main/packages/core/background/index.js#L1
          https://github.com/OctoLinker/OctoLinker/blob/main/babel.config.js
        - or specific file version
          https://github.com/OctoLinker/OctoLinker/blob/b97dfbfdbf3dee5f4836426e6dac6d6f473461db/.babelrc
          https://github.com/OctoLinker/OctoLinker/blob/cf7ec7de084da1c946c3450929d911800591e7e6/packages/core/background/index.js#L1
   2. FILE BLAME VIEW
      - IDentify by URL - Same as file view, but with `/blame/` instead of `/tree/` or `/blob/`
        - Either points to the latest commit of given branch
          https://github.com/OctoLinker/OctoLinker/blame/main/packages/core/background/index.js#L1
          https://github.com/OctoLinker/OctoLinker/blame/main/babel.config.js
        - Or specific file version
          https://github.com/OctoLinker/OctoLinker/blame/b97dfbfdbf3dee5f4836426e6dac6d6f473461db/.babelrc
          https://github.com/OctoLinker/OctoLinker/blame/cf7ec7de084da1c946c3450929d911800591e7e6/packages/core/background/index.js#L1
   3. EDIT VIEW
      - Identified by having `/edit/` in URL, and then selecting PREVIEW tab (a `<table>` will appear in HTML)
        https://github.com/OctoLinker/OctoLinker/edit/main/babel.config.js
      - NOTE: The ACTUAL edit view, where one is writing code, should NOT be highlighted.
   4. COMMIT
      - Identified by having `/commit/` in URL, e.g.
        https://github.com/OctoLinker/OctoLinker/commit/b97dfbfdbf3dee5f4836426e6dac6d6f473461db?diff=unified#diff-e56633f72ecc521128b3db6586074d2c
      4.1. Layout - Unified
        https://github.com/OctoLinker/OctoLinker/commit/b97dfbfdbf3dee5f4836426e6dac6d6f473461db?diff=unified#diff-e56633f72ecc521128b3db6586074d2c
      4.2. Layout - Split
        https://github.com/OctoLinker/OctoLinker/commit/b97dfbfdbf3dee5f4836426e6dac6d6f473461db?diff=split#diff-e56633f72ecc521128b3db6586074d2c
   5. PR DIFF VIEW (after PR was created)
       - Identified by PR URL:
         https://github.com/OctoLinker/OctoLinker/pull/873/files
       - TODO - DOES IT CHANGE IF THERE ARE COMMENTS?
   6. COMPARE VIEW (before PR is created)
       - Identified by `/compare/` in URL
         https://github.com/OctoLinker/OctoLinker/compare/main...fix-1165?diff=split&w
      6.1. Layout - Unified
        https://github.com/OctoLinker/OctoLinker/compare/main...fix-1165?diff=unified
      6.2. Layout - Split
        https://github.com/OctoLinker/OctoLinker/compare/main...fix-1165?diff=split
